### DISCUSSION TITLE
How to add fragments

### DISCUSSION THREAD

**Discussion by arunmenon1975** [hayes/pothos#392] [Q&A] [answered]
My `Prisma`-based project runs on a `Yoga` Graphql server and i just recently installed `Pothos` (with the `prisma plugin`) to get a better grip on the whole dev experience. Its been great so far. 

I am quite new to the whole stack really: graphql/yoga/prisma and i just came across Pothos last week.

My overall needs are quite simple.  But i just came across a need where i might have to use `fragments`.  I am so new to graphql that i have never used fragments even in the simple test creations i have attempted till now.  
 

I have a:

`sections > each_with_categories > each_with_maybe_some_subcategories > each_with_maybe some_more_subcategories > etc > items 
`

sort of UI need. The query could be like this:

```
sections{
    title
    categories {
       title
        items {
            title
        }
        categories {
            title
            items {
                title
            }
            categories {
               title
                items {
                    title
                }
                categories {
                    title
                    items {
                        title
                    }
                }
            }
        }
    }
}
```

One of the usecases is a collapsible tree in the UI.  

I have the relations in place in my Prisma model which obviously has some self-referencing relations for categories.

Ive gathered that of course there is no way to get an indefinite depth in my query and it probably might be best to have a hard-limit. Its more or less like the needs expressed [here](https://stackoverflow.com/questions/44746923/how-to-model-recursive-data-structures-in-graphql)  and [here](https://stackoverflow.com/questions/56314654/nested-query-to-unknown-level-graphql).  

I am ok with having a pre-defined depth. Querying the depths however will mean a very verbose GQL query. I think fragments can help here. 

I haven't found any mention of fragments in the `Pothos` docs. How do i go about defining fragments? If it indeed is possible, should this be done at a more root-level or can this be achieved at field-level?

**Comment by hayes** [comment, marked as answer]
Fragments are only used in queries.  All spec compliant graphql execution environments should support them.  Pothos only generates your graphql schema.  How queries are executed (including interpreting fragments) is entirely up to the library that executes the queries in your server.

Most JS graphql servers (yoga, apollo, etc) just use the official graphql library for executing queries, and have options for replacing the default execution with graphql-jit.  In either case, these fully support fragments.  So there is nothing you need to add or change in Pothos, or your server environment to be able to use fragments with your existing graphql schema.

A query using a fragment might look something like:

```graphql
query {
  sections {
    title
    categories {
      ...CategorySelection
      categories {
        ...CategorySelection
        categories {
          ...CategorySelection
          categories {
            ...CategorySelection
          }
        }
      }
    }
  }
}

fragment CategorySelection on Category {
  title
  items {
    title
  }
}
```

**Reply by arunmenon1975** [reply]
Thanks. I get it now - it's all cleared up. 

I think if i had tried fragments earlier in some form or the other i just might have understood the core concept earlier. Just maybe since Pothos seems to have it all covered i sort of took it for granted that there could be something for fragments as well. 

It's been a great experience using Pothos so far. And the prisma plugin (which incidentally i saw only much later, that also has the n+1 mitigation built-in) integration gives it the end-to-end full-typings support and can potentially help ward of lots of potential bugs while developing.
